### PR TITLE
fixing ZipPackagePart.GetStreamCore crashes with NotSupportedException

### LINKED ISSUE #39816: ZipPackagePart.GetStreamCore crashes with NotSupportedException
This spawned off investigations done in https://github.com/dotnet/wpf/issues/1363.

A simple reproduction in a console application can be obtained by doing:

``` 
var zp = ZipPackage.Open("test.zip", System.IO.FileMode.Create, System.IO.FileAccess.Write);
var part = zp.CreatePart(new System.Uri("/test.txt", UriKind.Relative), "");
var stream = part.GetStream(System.IO.FileMode.Create);
```

A NotSupportedException will arise due to this code:

https://github.com/dotnet/corefx/blob/a10890f4ffe0fadf090c922578ba0e606ebdd16c/src/System.IO.Packaging/src/System/IO/Packaging/ZipPackagePart.cs#L30-L36

The call to `SetLength` fails since the underlying stream does not support this.

The flow here is:
* https://github.com/dotnet/corefx/blob/78589e4d2c98bf71cefa4dbc94cde3783b60a934/src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs#L290
* `ZipArchiveEntry.GetDataCompressor` creates a new `DeflateStream` which is then wrapped, along with the original stream, by `CheckSumAndSizeWriteStream`.
https://github.com/dotnet/corefx/blob/78589e4d2c98bf71cefa4dbc94cde3783b60a934/src/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs#L600
* `CheckSumAndSizeWriteStream` is never seekable and doesn't support `SetPosition`:
https://github.com/dotnet/corefx/blob/a10890f4ffe0fadf090c922578ba0e606ebdd16c/src/System.IO.Compression/src/System/IO/Compression/ZipCustomStreams.cs#L364

It seems that the assumption about the underlying stream is incorrect or the underlying stream should be allowing for this functionality.

### PR DESCRIPTION
**Pull request by stevenbrix**
The exception thrown here by `ZipPackagePart` is dubious and unnecessary, and an unintended mistake from the initial port of `System.IO.Packaging` from WindowsBase. ZipArchiveEntry only ever supports opening once when the backing archive is in Create mode, so the call to SetLength is unnecessary. You could still open an archive in Update mode then call part.GetStream(FileMode.Create), in which case we'll want this call to SetLength, so we only avoid this call when the backing Archive is in Create mode.

Fixes #39816

### CODE REVIEW

File: src/System.IO.Packaging/tests/Tests.cs [hunk 1]
```diff
@@ -3892,7 +3892,7 @@ public void ComparePackUriDifferentPack()
         }
 
         [Fact]
-        void CreatePackUriWithFragment()
+        public void CreatePackUriWithFragment()
```
Comment (stephentoub) [inline src/System.IO.Packaging/tests/Tests.cs]: This is surprising.  We have analyzers that I thought should have caught this.
cc: @ViktorHofer

Comment (ViktorHofer) [reply src/System.IO.Packaging/tests/Tests.cs]: analyzers don't run for test projects...

Comment (stephentoub) [reply src/System.IO.Packaging/tests/Tests.cs]: > analyzers don't run for test projects...

xunit analyzers _only_ run for test projects.

Comment (ViktorHofer) [reply src/System.IO.Packaging/tests/Tests.cs]: https://github.com/dotnet/corefx/blob/0b822c65aa586d9a62239ee5c862fede2543a443/src/Directory.Build.props#L52

Comment (stephentoub) [reply src/System.IO.Packaging/tests/Tests.cs]: @ViktorHofer, I'm not sure what you're trying to show me with that Directory.Build.props line.  Why wouldn't xunit analyzers run on this xunit test project?

Comment (ViktorHofer) [reply src/System.IO.Packaging/tests/Tests.cs]: Oh I thought you were talking about the roslyn code-rule that forces access modifiers to be specified.

Comment (ViktorHofer) [reply src/System.IO.Packaging/tests/Tests.cs]: There is no xunit analyzer that specifies that test methods must be public: https://xunit.net/xunit.analyzers/rules/

Comment (stephentoub) [reply src/System.IO.Packaging/tests/Tests.cs]: > There is no xunit analyzer that specifies that test methods must be public

Ugh.  That's unfortunate.  Thanks.

Comment (ViktorHofer) [reply src/System.IO.Packaging/tests/Tests.cs]: >  I'm not sure what you're trying to show me with that Directory.Build.props line. Why wouldn't xunit analyzers run on this xunit test project?

I was talking about the analyzers specified here: https://github.com/dotnet/corefx/blob/master/eng/Analyzers.props. Of course xunit analyzers run on test projects but there is no rule that forces a method to be public. The roslyn analyzers would have caught that.

Comment (stephentoub) [reply src/System.IO.Packaging/tests/Tests.cs]: > The roslyn analyzers would have caught that.

Sort of.  They'd have flagged it, but then an auto-fix would have explicitly added "private".

**Comment by stevenbrix** [comment]
@stephentoub some tests are failing in UWP due to the file system restrictions, i tried following the instructions [here](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/developer-guide.md#building-tests-with-uwp-windows-only) but run into these errors locally:

>C:\Program Files\dotnet\sdk\3.0.100-preview9-013756\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ConflictResolution.targets(39,5): error NETSDK1042: Could not load PlatformManifest from 'D:\dev\src\dotnet\corefx\artifacts\bin\testhost\uap-Windows_NT-Debug-x64\UAPLayout\PlatformManifest.txt' because it did not exist. [D:\dev\src\dotnet\corefx\src\Common\tests\CoreFx.Private.TestUtilities\CoreFx.Private.TestUtilities.csproj]

Is there something else i need to do?

**Comment by stephentoub** [comment]
@ViktorHofer, can you help @stevenbrix?

**Comment by ViktorHofer** [comment]
Sure. You need to build from root for uap first : build -f uap.

**Pull request merged**